src/main/java/org/dsa/iot/alarm/AbstractAlarmObject.java

```diff
@@ -0,0 +1,322 @@
+            }
+        } else {
+            System.out.println(getClass().getName());//XXX
+        }
+        initProperties();
+        initActions();
+    }
+
+    /**
+     * Subclass hook for adding actions, called by init(node) after the subtree has been
+     * loaded.
+     */
+    protected void initActions() {
+    }
+
+    /**
+     * Subclass hook for adding configs and attributes, called by init(node) after the
+     * subtree has been loaded.
+     */
+    protected void initProperties() {
+    }
+
+    /**
+     * Returns the value of the enabled property.  If there is no enabled
+     * property, will return true;
+     */
+    public boolean isEnabled() {
+        Value val = getProperty(ENABLED);
+        if (val == null) {
+            return true;
+        }
+        return val.getBool().booleanValue();
+    }
+
+    @Override public final boolean isSteady() {
+        return steady;
+    }
+
+    /**
+     * A convenience for isEnabled() && isSteady()
+     */
+    public boolean isValid() {
+        if (!isEnabled() || isSteady()) {
+            return false;
+        }
+        /*
+        AlarmObject parent = getParent();
+        if (parent instanceof AbstractAlarmObject) {
+            if (!((AbstractAlarmObject)parent).isValid()) {
+                return false;
+            }
+        }
+        */
+        return true;
+    }
+
+    /**
+     * A convenience that creates and adds a new AlarmObject as well as the corresponding
+     * persistent SDK Node.
+     *
+     * @param name            The name of the SDK node.
+     * @param alarmObjectType The alarm object class.
+     * @return The newly created and added child.
+     * @throws IllegalArgumentException If a child with the same name already exists.
+     * @throws RuntimeException         Wrapping any other non-runtime exceptions.
+     */
+    public AlarmObject newChild(String name, Class alarmObjectType) {
+        AlarmObject ret = null;
+        try {
+            Node parent = getNode().getParent();
+            Node node = parent.getChild(name);
+            if (node != null) {
+                throw new IllegalArgumentException("Name already in use: " + name);
+            }
+            node = parent.createChild(name).setSerializable(true).build();
+            ret = (AlarmObject) alarmObjectType.newInstance();
+            ret.init(node);
+            addChild(ret);
+        } catch (Exception x) {
+            AlarmUtil.throwRuntime(x);
+        }
+        return ret;
+    }
+
+    /**
+     * Will remove all descendant alarm objects thus they will get the removed callback
+     * as well.
+     * <p/>{@inheritDoc}
+     */
+    @Override public void removed() {
+        removeAllDescendants();
+        node.getParent().removeChild(node);
+    }
+
+    /**
+     * A convenience that calls removeChild for all direct descendants.
+     */
+    protected void removeAllDescendants() {
+        for (int i = childCount(); --i >= 0; ) {
```

Comment:
The compiler won't optimize removal from the collection.  This take the tail end of the array list which won't result in an array copy to shift the remaining elements.  I always try to access random access collections in reverse, there can be a number of benefits.